src/iocp/mod.rs

```diff
@@ -0,0 +1,821 @@
+                        new_events += 1;
+                    }
+                    FeedEventResult::Notified => {
+                        notified = true;
+                    }
+                }
+            }
+
+            // Break if there was a notification or at least one event, or if deadline is reached.
+            let timeout_is_empty =
+                timeout.map_or(false, |t| t.as_secs() == 0 && t.subsec_nanos() == 0);
+            if notified || new_events > 0 || timeout_is_empty {
+                break;
+            }
+
+            log::trace!("wait: no events found, re-entering polling loop");
+        }
+
+        Ok(())
+    }
+
+    /// Notify this poller.
+    pub(super) fn notify(&self) -> io::Result<()> {
+        // Push the notify packet into the IOCP.
+        self.port.post(0, 0, self.notifier.clone())
+    }
+
+    /// Run an update on a packet.
+    fn update_packet(&self, mut packet: Packet) -> io::Result<()> {
+        loop {
+            // If we are currently polling, we need to update the packet immediately.
+            if self.polling.load(Ordering::Acquire) {
+                packet.update()?;
+                return Ok(());
+            }
+
+            // Try to queue the update.
+            match self.pending_updates.push(packet) {
+                Ok(()) => return Ok(()),
+                Err(p) => packet = p.into_inner(),
+            }
+
+            // If we failed to queue the update, we need to drain the queue first.
+            self.drain_update_queue(true)?;
+        }
+    }
+
+    /// Drain the update queue.
+    fn drain_update_queue(&self, limit: bool) -> io::Result<()> {
+        let max = if limit {
+            self.pending_updates.capacity().unwrap()
+        } else {
+            std::usize::MAX
+        };
+
+        // Only drain the queue's capacity, since this could in theory run forever.
+        for _ in 0..max {
+            if let Ok(packet) = self.pending_updates.pop() {
+                packet.update()?;
+            } else {
+                return Ok(());
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Get a handle to the AFD reference.
+    fn afd_handle(&self) -> io::Result<Arc<Afd<Packet>>> {
+        const AFD_MAX_SIZE: usize = 32;
+
+        // See if there are any existing AFD instances that we can use.
+        let mut afd_handles = self.afd.lock().unwrap_or_else(|e| e.into_inner());
+        if let Some(handle) = afd_handles.iter().find(|h| {
+            let ref_count = Arc::strong_count(h).saturating_sub(1);
+            ref_count < AFD_MAX_SIZE
+        }) {
+            return Ok(handle.clone());
+        }
+
+        // Create a new AFD instance.
+        let afd = Arc::new(Afd::new()?);
+
+        // Register the AFD instance with the I/O completion port.
+        self.port.register(&*afd, true)?;
+
+        // Insert a copy of the AFD instance into the list.
+        afd_handles.push(afd.clone());
```

Comment:
do we ever prune AFDs from `afd_handles`? could this exhaust all available memory? There currently doesn't seem to be a way to tell the poller that it should `shrink_to_fit`, free all cached blocks that aren't used anymore (or, if I understand the usage of `afd_handles` correctly, it seems to act as a cache and we could get away with
```rust
self.afd = Mutex::new(Vec::new());
```

Reply:
Each AFD handle can be associated with up to 32 sockets. We use the ref-counting mechanism in `Arc` to keep track of how many sockets are being used. This ensures that old AFDs are reused by old sockets. I guess you could end up with thousands of AFD handles if you open and close 100,000's of sockets. In that case, I can add a pruning mechanism for when there are a large number of AFD handles with no associated sockets.

Reply:
alternatively, you could "just" use weak pointers, that way, opening, polling, and then closing 100'000's of sockets should automatically free the associated AFDs, and the next crawl of the list could then prune the gone ones.